File: core/src/passes/redundant_load.cpp
// core/src/passes/redundant_load.cpp — F1-PS-007.
//
// Re-reading the same memory location through the same address ref
// with no intervening memory write is redundant. We rewrite the second
// LoadMem to `Or %v1, %v1` (our copy idiom) so copy_propagate + DCE
// handle the downstream cleanup uniformly.
//
// Scope decisions:
//   * LoadMemTSO is untouched. Acquire semantics make a second load a
//     real synchronisation event whose elimination would change
//     observable behaviour even on single-threaded code under the
//     adaptive TSO pass.
//   * Any StoreMem / StoreMemTSO / fence-like op flushes the whole
//     load table — we have no alias analysis, so any write is assumed
//     to possibly alias every tracked address.
//   * Size matters: `LoadMem %a, I32` and `LoadMem %a, I64` describe
//     different values at the same address; only `(addr, size)` tuples
//     dedupe.

#include "prisma/passes.hpp"

#include <unordered_map>
#include <utility>
#include <variant>

namespace prisma::passes {

namespace {

struct LoadKey {
    ir::Ref    addr;
    ir::OpSize size;
    bool operator==(const LoadKey&) const = default;
};

struct LoadKeyHash {
    std::size_t operator()(const LoadKey& k) const noexcept {
        return std::hash<ir::Ref>{}(k.addr) * 31u
             + static_cast<std::size_t>(k.size);
    }
};

}  // namespace

std::vector<ir::Stmt>
redundant_load_eliminate(const std::vector<ir::Stmt>& stmts) {
    std::unordered_map<LoadKey, ir::Ref, LoadKeyHash> last_load;

    std::vector<ir::Stmt> out;
    out.reserve(stmts.size());

    for (const auto& s : stmts) {
        if (std::holds_alternative<ir::LoadMem>(s.op) && s.result) {
            const auto& l = std::get<ir::LoadMem>(s.op);
            const LoadKey k{l.addr, l.size};
            auto it = last_load.find(k);
            if (it != last_load.end()) {
                // Rewrite as a copy of the prior load's result.
                ir::Stmt rewritten{s.result,
                    ir::Op{ir::BinOp{
                        ir::BinOpKind::Or, it->second, it->second, l.size}}};
                out.push_back(std::move(rewritten));
                continue;
            }
            last_load[k] = *s.result;
        }
        else if (std::holds_alternative<ir::StoreMem>(s.op)
              || std::holds_alternative<ir::StoreMemTSO>(s.op)
              || std::holds_alternative<ir::Fence>(s.op)) {
            // Any store could alias any tracked address. A fence can make
            // surrounding memory operations observable, so do not forward
            // plain loads across it.
            last_load.clear();
        }
        // LoadMemTSO, Jumps, Calls, CmpFlags, pure ops — no effect on the
        // memory image visible via plain LoadMem, so the table survives.

        out.push_back(s);
    }

    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/redundant_load.rs
//! Redundant load elimination pass.
//!
//! Mirrors C++ `redundant_load_eliminate`. A second `LoadMem` through the
//! same `(addr_ref, size)` with no intervening store/fence is rewritten to a
//! copy of the prior load's result (`Or %v, %v`), letting copy-prop + DCE do
//! downstream cleanup. `LoadMemTSO` is never forwarded (acquire semantics).

use std::collections::HashMap;

use prisma_ir::{BasicBlock, BinOp, BinOpKind, Function, Op, OpSize, Ref, Stmt};

use crate::Pass;

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct RedundantLoad;

impl RedundantLoad {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for RedundantLoad {
    fn name(&self) -> &'static str {
        "redundant_load_eliminate"
    }

    fn run(&self, func: Function) -> Function {
        redundant_load_eliminate(func)
    }
}

/// Per-block redundant-load elimination.
pub fn redundant_load_eliminate(func: Function) -> Function {
    let blocks = func
        .blocks
        .into_iter()
        .map(|block| {
            // key: (addr_ref, size) -> prior load result ref
            let mut last_load: HashMap<(Ref, OpSize), Ref> = HashMap::new();
            let mut out = Vec::with_capacity(block.stmts.len());

            for stmt in block.stmts {
                match &stmt.op {
                    Op::LoadMem(l) => {
                        if let Some(result) = stmt.result {
                            let key = (l.addr, l.size);
                            if let Some(&prev) = last_load.get(&key) {
                                out.push(Stmt::new(
                                    Some(result),
                                    Op::BinOp(BinOp {
                                        op: BinOpKind::Or,
                                        lhs: prev,
                                        rhs: prev,
                                        size: l.size,
                                    }),
                                ));
                                continue;
                            }
                            last_load.insert(key, result);
                        }
                    }
                    Op::StoreMem(_) | Op::StoreMemTSO(_) | Op::Fence(_) => {
                        // Any store could alias; a fence orders memory.
                        last_load.clear();
                    }
                    _ => {}
                }
                out.push(stmt);
            }

            BasicBlock {
                id: block.id,
                stmts: out,
            }
        })
        .collect();

    Function {
        blocks,
        entry: func.entry,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{Fence, FenceKind, LoadMem, StoreMem};

    fn load(result: u32, addr: u32, size: OpSize) -> Stmt {
        Stmt::new(Some(result), Op::LoadMem(LoadMem { addr, size }))
    }

    #[test]
    fn second_identical_load_becomes_copy() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![load(0, 5, OpSize::I64), load(1, 5, OpSize::I64)],
            }],
        };
        let out = redundant_load_eliminate(func);
        match &out.blocks[0].stmts[1].op {
            Op::BinOp(b) => {
                assert_eq!(b.op, BinOpKind::Or);
                assert_eq!(b.lhs, 0);
                assert_eq!(b.rhs, 0);
            }
            other => panic!("expected copy, got {other:?}"),
        }
    }

    #[test]
    fn store_between_loads_blocks_forwarding() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    load(0, 5, OpSize::I64),
                    Stmt::new(
                        None,
                        Op::StoreMem(StoreMem {
                            addr: 6,
                            value: 0,
                            size: OpSize::I64,
                        }),
                    ),
                    load(1, 5, OpSize::I64),
                ],
            }],
        };
        let out = redundant_load_eliminate(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn fence_blocks_forwarding() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    load(0, 5, OpSize::I32),
                    Stmt::new(
                        None,
                        Op::Fence(Fence {
                            kind: FenceKind::Mfence,
                        }),
                    ),
                    load(1, 5, OpSize::I32),
                ],
            }],
        };
        let out = redundant_load_eliminate(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn different_size_is_not_redundant() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![load(0, 5, OpSize::I32), load(1, 5, OpSize::I64)],
            }],
        };
        let out = redundant_load_eliminate(func.clone());
        assert_eq!(out, func);
    }
}
